### PR TITLE
Fix for magics and shell escapes

### PR DESCRIPTION
**Pull request by rchiodo**
Check front of diagnostic for % or !. If this is the first char on a line, filter out the diag

### CODE REVIEW

File: src/notebookConverter.ts [hunk 1]
```diff
@@ -47,422 +51,497 @@ function toPosition(positionLike: Position): Position {
...
+          cellUris.push(document.uri.toString());
+        });
+      // Possible some cells were deleted, we need to clear the diagnostics of those cells as well.
+      const currentCellUris = new Set(cellUris);
+      oldCellUris
+        .filter((cellUri) => !currentCellUris.has(cellUri))
+        .forEach((cellUri) => result.set(Uri.parse(cellUri), []));
+      this.mapOfConcatDocumentsWithCellUris.set(uri.toString(), cellUris);
+
+      // Filter out any diagnostics that have to do with magics or shell escapes
```
Comment (rchiodo) [inline src/notebookConverter.ts:186]: This is the actual bug fix

File: src/notebookConverter.ts [hunk 2]
```diff
@@ -1,40 +1,44 @@
...
-    Uri,
-    WorkspaceEdit,
-    NotebookDocument
-} from 'vscode';
-import { IVSCodeNotebook } from './common/types';
-import { InteractiveInputScheme, InteractiveScheme, NotebookCellScheme } from './common/utils';
-import { IFileSystem } from './common/types';
-import { IConcatTextDocument } from './concatTextDocument';
-import { NotebookConcatDocument } from './notebookConcatDocument';
+  CodeAction,
```
Comment (rebornix) [inline src/notebookConverter.ts:5]: did the indentation change from 4 spaces to 2 spaces?

File: src/notebookConverter.ts [hunk 3]
```diff
@@ -511,203 +590,264 @@ export class NotebookConverter implements Disposable {
...
+
+  private filterMagics(
+    wrapper: NotebookConcatDocument,
+    value: Diagnostic
+  ): boolean {
+    // Get the code at the range
+    const text = wrapper.getText(value.range);
+
+    // Only skip diagnostics on the front of the line (spacing?)
+    if (
```
Comment (rebornix) [inline src/notebookConverter.ts:844]: e.g. `/^\s*\%/` ?

**Review by rebornix — APPROVED** [review]
Maybe we can add indentation setting to the workspace so we don't reindent whole document next time. The fix looks good to me.

File: src/notebookConverter.ts (hunk 2 above)
Comment (rchiodo) [reply src/notebookConverter.ts]: Not sure what's doing that? Prettier is my default formatter. Looks like it did? I'll add a .prettierrc.js

File: src/notebookConverter.ts (hunk 3 above)
Comment (rchiodo) [reply src/notebookConverter.ts]: No I don't think we want to remove ones that aren't at the front of the line.

**Review by IanMatthewHuff — APPROVED** [review]

**Comment by ryan-feeley** [comment]
It appears to me that this issue/bug still exits for `.ipy` files. I.e., "IPython script files". For this file type, I believe pylance should not produce diagnostic errors on magics.

**Comment by rchiodo** [comment]
@ryan-feeley this repository is specifically for notebook support. Supporting magics in ```.ipy``` would be an issue on pylance itself.

**Comment by cwstanton** [comment]
Not sure the bug fix, as described above, really satisfies the entirety of the need.  The following is a perfectly valid Jupyter Notebook/Python Kernel cell, but not in VSCode:

%load_ext sql
%sql sqlite:///sqlite/r2020d2021.db

%%sql SELECT
    abscode.type,
    sum(abssum.actval) as SumActVal,
    sum(abssum.assdval) as SumAssdVal
FROM
    abstract_summary_txt as abssum
INNER JOIN
    abscode_csv as abscode
    ON abssum.propcd = abscode.propcd
WHERE
    abscode.propcd NOT LIKE '9%'
GROUP BY
    abscode.type

**Comment by rchiodo** [comment]
@cwstanton yes other languages are out of scope for this fix. It only fixes the %% and ! diagnostic errors. You'd likely get this:

image

It's likely going to be a while before we dynamically switch the language on a cell and have pylance ignore it.

**Comment by cwstanton** [comment]
It would be great if there were a "brute force" way to just turn off pylance for a cell.  I can imagine something like a cell's top line:

# pylance: ignore-cell

**Comment by ryan-feeley** [comment]
> 
> 
> @ryan-feeley this repository is specifically for notebook support. Supporting magics in `.ipy` would be an issue on pylance itself.

Thanks @rchiodo . I've started a feature request [discussion on the pylance board](https://github.com/microsoft/pylance-release/discussions/1868). We'll see where it goes. Take care!

**Comment by starball5** [comment]
Related Stack Overflow question: [VS Code problem when using magic commands in Jupyter notebooks: "Expected expression Pylance"](https://stackoverflow.com/q/71546251/11107541)

**Pull request merged**